frontend/src/Frontend/UI/Dialogs/AddVanityAccount/DefineKeyset.hs

```diff
@@ -0,0 +1,173 @@
+        (\k t -> IntMap.singleton k . fmap keyToText <$> _externalKeyInput_value t)
+
+  v <- uiAdditiveInput
+    (const uiPubkeyInput)
+    _externalKeyInput_input
+    (not . T.null)
+    T.null
+    T.empty
+    (current preselections <@ pb)
+
+  pure $ KeysetInputs v (v >>= toSet)
+
+defineKeyset
+  :: forall t m key model
+     . ( MonadWidget t m
+       , HasWallet model key t
+       )
+  => model
+  -> Dynamic t (IntMap (Dropdown t (Maybe Int)))
+  -> m (KeysetInputs t (Dropdown t (Maybe Int)) PublicKey)
+defineKeyset model preselections0 = do
+  let
+    selectMsgKey = Nothing
+    selectMsgMap = Map.singleton selectMsgKey "Select"
+
+    keyElem k v = Map.singleton (Just k) $ keyToText $ _keyPair_publicKey $ _key_pair v
+
+    dAllKeys = mappend selectMsgMap . IntMap.foldMapWithKey keyElem <$> model ^. wallet_keys
+
+    uiSelectKey k = mkLabeledClsInput False (constDyn T.empty) $ const
+      $ uiDropdown k dAllKeys $ def
+      & dropdownConfig_attributes .~ constDyn ("class" =: "labeled-input__input")
+
+    toIntSet :: IntMap.IntMap (Dropdown t (Maybe Int)) -> Dynamic t IntSet.IntSet
+    toIntSet = fmap (IntSet.fromList . IntMap.elems) . wither value
+
+    preselections = preselections0 >>= IntMap.foldMapWithKey
+      (\k dd -> IntMap.singleton k . Just <$> _dropdown_value dd)
+
+  pb <- getPostBuild
+
+  dSelectedKeys <- uiAdditiveInput
+    (const uiSelectKey)
+    _dropdown_change
+    (/= selectMsgKey)
+    (== selectMsgKey)
+    selectMsgKey
+    (current preselections <@ pb)
+
+  pure $ KeysetInputs dSelectedKeys $ ffor2 (model ^. wallet_keys) (dSelectedKeys >>= toIntSet) $ \wKeys ->
+    Set.fromDistinctAscList . IntMap.elems . fmap (_keyPair_publicKey . _key_pair) . IntMap.restrictKeys wKeys
+
+uiDefineKeyset
+  :: ( MonadWidget t m
+     , HasWallet model key t
+     , HasJsonData model t
+     )
+  => model
+  -> DefinedKeyset t
+  -> m (Dynamic t (Maybe AddressKeyset), DefinedKeyset t)
+uiDefineKeyset model presets = do
+  pb <- getPostBuild
+  let
+    allPreds = fmap (catMaybes . Map.elems)
+      $ joinDynThroughMap
+      $ fmap _keyset_pred
+      <$> model ^. jsonData_keysets
+
+    allPredSelectMap = ffor allPreds $ \ps ->
+      Map.fromList . fmap (\x -> (x,x)) $ ps <> predefinedPreds
+
+  rec
+    selectedKeys <- mkLabeledClsInput False "Chainweaver Keys" $ const
+      $ defineKeyset model $ _keysetInputs_value $ _definedKeyset_internalKeys presets
+
+    externalKeys <- mkLabeledClsInput False "External Keys" $ const
+      $ uiExternalKeyInput $ _keysetInputs_value $ _definedKeyset_externalKeys presets
+
+    predicate <- mkLabeledClsInput False "Predicate (Keys Required to Sign for Account)" $ const
+      $ fmap value $ uiDropdown mempty allPredSelectMap $ def
+      & dropdownConfig_attributes .~ constDyn ("class" =: "labeled-input__input")
+      & dropdownConfig_setValue .~ (current (_definedKeyset_predicate presets) <@ pb)
+
+  -- TODO validate this (?? validation ??)
```

Comment:
This comment may be obsolete now. I originally made it for the freeform text predicate